### PR TITLE
update `hash_stable` for `List<Ty<'tcx>>`

### PR DESCRIPTION
**Pull request by lcnr**
cc https://github.com/rust-lang/rust/pull/93505#issuecomment-1047538798

this is the hottest part changed since the pre-merge perf run

### CODE REVIEW

File: compiler/rustc_middle/src/ty/impls_ty.rs [hunk 1]
```diff
@@ -73,18 +73,16 @@ impl<'a, 'tcx> HashStable<StableHashingContext<'a>> for ty::subst::GenericArgKin
...
-            // their discriminant (since the discriminant of `TyKind` is unlikely to ever start
-            // with 0xFF).
+            // hashed for one of the other variants, we hash some very high number instead
+            // of their actual discriminant since `TyKind` should never start with anything
+            // that high.
             ty::subst::GenericArgKind::Type(ty) => ty.hash_stable(hcx, hasher),
             ty::subst::GenericArgKind::Const(ct) => {
-                0xFFu8.hash_stable(hcx, hasher);
-                mem::discriminant(self).hash_stable(hcx, hasher);
+                0xF3u8.hash_stable(hcx, hasher).hash_stable(hcx, hasher);
```
Comment (Kobzol) [inline compiler/rustc_middle/src/ty/impls_ty.rs:81]: Is this supposed to be hashed twice? :thinking:

Comment (lcnr) [reply compiler/rustc_middle/src/ty/impls_ty.rs:81]: definitely not, i guess that hash stable is a noop :sweat_smile:
```suggestion
                0xF3u8.hash_stable(hcx, hasher);
```

**Comment by lcnr** [comment]
will probably have to try something different

@bors try @rust-timer queue

**Comment by lcnr** [comment]
alright, wasn't the reason for the perf impact, but I also believe that this is a small improvement. If this is merged nobody else will try the same change in the future and it doesn't make anything worse :shrug:

@bors ready

**Comment by petrochenkov** [comment]
I don't understand what was the point of https://github.com/rust-lang/rust/pull/93505.
It was supposed to be an optimization (at cost of some extra complexity), but it actually wasn't. Shouldn't it be reverted entirely?

**Comment by petrochenkov** [comment]
I also don't understand how the `HashStable` works and why it's correct.
In case of enums like
```rust
enum E {
  A(u8),
  B(u8),
}
```
hashing the discriminant (or some other distinct number) is necessary because the payloads have the same type and may collide.

But with
```rust
enum E {
  A(OneType),
  B(EntirelyDifferentType),
}
```
it's not clear why the discriminant should be hashed, and why hashing `OneType + discriminant` and `EntirelyDifferentType + discriminant` has any lower chance of collision than `OneType` and `EntirelyDifferentType`.

**Comment by lcnr** [comment]
> I don't understand what was the point of #93505. It was supposed to be an optimization (at cost of some extra complexity), but it actually wasn't. Shouldn't it be reverted entirely?

I mostly added this to enable `ty::Tuple` to use `List<Ty<'tcx>>` instead of `List<GenericArg<'tcx>>`. Going to try to keep the `List<Ty<'tcx>>` and `List<GenericArg<'tcx>>` split in a separate PR to check the impact of that.

> I also don't understand how the `HashStable` works and why it's correct. In case of enums like
> 
> ```rust
> enum E {
>   A(u8),
>   B(u8),
> }
> ```
> 
> hashing the discriminant (or some other distinct number) is necessary because the payloads have the same type and may collide.
> 
> But with
> 
> ```rust
> enum E {
>   A(OneType),
>   B(EntirelyDifferentType),
> }
> ```
> 
> it's not clear why the discriminant should be hashed, and why hashing `OneType + discriminant` and `EntirelyDifferentType + discriminant` has any lower chance of collision than `OneType` and `EntirelyDifferentType`.

if you have
```rust
enum OneType {
    VariantA(u32),
    VariantB(u8),
}

enum EntirelyDifferentType {
    VariantA,
    VariantB(u8),
}
```
hashing `E` without considering its discriminant would result in `E::A(OneType::VariantB(3))` and `E::B(EntirelyDifferentType::VariantB(3))` to have the same hash.

I don't think it is trivially apparent that `ty::RegionKind` and `ty::ConstKind` do not have a same-ish variants with the same discriminant.

**Comment by petrochenkov** [comment]
I'll mark this as blocked on https://github.com/rust-lang/rust/pull/94841 for now.
@rustbot blocked

**Comment by lcnr** [comment]
keeping the reasoning from https://github.com/rust-lang/rust/pull/94799#issuecomment-1064309755

> alright, wasn't the reason for the perf impact, but I still believe that this is a small improvement. If this is merged nobody else will try the same change in the future and it doesn't make anything worse :shrug:

@petrochenkov depending on your opinion here, feel free to either approve or close this PR
@bors ready

**Pull request merged**